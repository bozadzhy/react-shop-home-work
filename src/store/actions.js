import { createAction } from "@reduxjs/toolkit";

import { sendRequest } from "../helpers/sendRequst";

export const actionAddProducts = createAction("ACTION_ADD_PRODUCTS");
export const actionAddToFavotites= createAction("ACTION_ADD_TO_FAVOTITES");
export const actionAddToCart= createAction("ACTION_ADD_TO_CART");
export const actionHandelModal= createAction("ACTION_HANDEL_MODAL");
export const actionMinusFromCart= createAction("ACTION_MINUS_FROM_CART");
export const actionClearCart= createAction("ACTION_CLEAR_CART");
export const actionDeleteItem = createAction("ACTION_DELETE_ITEM");


export const actionFetchProducts = () => (dispatch) => {
  return (sendRequest('products.json').then((results) => {
    dispatch(actionAddProducts(results));
  }))
};

