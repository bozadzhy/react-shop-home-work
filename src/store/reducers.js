import { createReducer } from "@reduxjs/toolkit";
import * as actions from "./actions.js";

const initialState = {
  products: [],
  favorites: JSON.parse(localStorage.getItem("favorite")) || [],
  cart: JSON.parse(localStorage.getItem("cart")) || [],
  isModal: false,
  totalPrice: JSON.parse(localStorage.getItem("totalPrice")) || 0,
};

export default createReducer(initialState, {
  [actions.actionAddProducts]: (state, { payload }) => {
    state.products = payload;
  },

  [actions.actionAddToFavotites]: (state, { payload }) => {
    const findObj = state.favorites.find((obj) => obj.id === payload.id);
    if (!findObj || findObj.length === 0) {
      localStorage.setItem(
        "favorite",
        JSON.stringify([...state.favorites, payload])
      );
      state.favorites = [...state.favorites, payload];
    } else {
      const filterFevorites = state.favorites.filter(
        (obj) => obj.id !== payload.id
      );
      localStorage.setItem("favorite", JSON.stringify(filterFevorites));
      state.favorites = [...filterFevorites];
    }
  },

  [actions.actionAddToCart]: (state, { payload }) => {
    const findCart = state.cart.find((obj) => obj.id === payload.id);

    if (!findCart || findCart.length === 0) {
      state.cart = [...state.cart, { ...payload, count: 1 }];

      state.totalPrice = state.cart.reduce((sum, obj) => {
        return obj.price * obj.count + sum;
      }, 0);
    } else {
      findCart.count++;
      state.totalPrice = state.cart.reduce((sum, obj) => {
        return obj.price * obj.count + sum;
      }, 0);
    }
  },

  [actions.actionMinusFromCart]: (state, { payload }) => {
    const findItem = state.cart.find((obj) => obj.id === payload.id);
    findItem.count--;
    if (findItem.count < 1) {
      findItem.count = 1;
    }
    state.totalPrice = state.totalPrice - findItem.price;
  },

  [actions.actionDeleteItem]: (state, { payload }) => {
    const findItem = state.cart.find((obj) => obj.id === payload.id);

    state.totalPrice = state.totalPrice - findItem.price * findItem.count;
    findItem.count--;

    const filterArr = state.cart.filter((obj) => obj.id !== payload.id);
    localStorage.setItem("cart", JSON.stringify(filterArr));
    state.cart = [...filterArr];
  },

  [actions.actionHandelModal]: (state) => {
    state.isModal = !state.isModal;
  },
  [actions.actionClearCart]: (state) => {
    state.cart = [];
    state.totalPrice = 0;
  },
});
