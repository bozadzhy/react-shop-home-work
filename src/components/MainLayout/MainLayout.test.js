import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import MainLayout from "./MainLayout";
import { BrowserRouter } from "react-router-dom";
import store from "../../store";
import NewThemeProvider from "../../context/context";

describe("MainLayout Snapshot Test", () => {
  it("should match snapshot for MainLayout component", () => {
    const component = renderer.create(
      <NewThemeProvider>
        <Provider store={store}>
          <BrowserRouter>
            <MainLayout />
          </BrowserRouter>
        </Provider>
      </NewThemeProvider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
