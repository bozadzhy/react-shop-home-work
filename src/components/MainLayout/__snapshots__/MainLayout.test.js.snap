// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainLayout Snapshot Test should match snapshot for MainLayout component 1`] = `
Array [
  <div
    className="root"
  >
    <div
      className="container"
    >
      <div>
        <a
          href="/"
          onClick={[Function]}
        >
          <svg
            className="logo"
          >
            logo.svg
          </svg>
        </a>
      </div>
      <div
        className="wrap"
      >
        <span
          className="theme"
          onClick={[Function]}
        >
          <svg>
            square1.svg
          </svg>
        </span>
        <a
          className="header-icon"
          href="/favorites"
          onClick={[Function]}
        >
          <svg>
            starIcon.svg
          </svg>
          <span>
            0
          </span>
        </a>
        <a
          className="header-icon"
          href="/cart"
          onClick={[Function]}
        >
          <svg>
            cart.svg
          </svg>
          <span>
            0
          </span>
        </a>
      </div>
    </div>
  </div>,
  <div
    className="footer"
  >
    Footer
  </div>,
]
`;
